packages/firebase_database/lib/src/database_reference.dart

```diff
@@ -125,10 +128,65 @@ class DatabaseReference extends Query {
   ///
   /// remove() is equivalent to calling set(null)
   Future<Null> remove() => set(null);
+
+  /// Determine the next transaction key to be used. 0 is used if there are
+  /// no existing keys, otherwise the last key is incremented by 1.
+  int _getNextTransactionKey() {
+    if (FirebaseDatabase._transactions.isEmpty) {
+      return 0;
+    }
+    final int lastKey = FirebaseDatabase._transactions.keys.last;
+    return lastKey + 1;
+  }
+
+  /// Performs an optimistic-concurrency transactional update to the data at
+  /// this Firebase Database location.
+  Future<Null> runTransaction(TransactionHandler transactionHandler,
```

Comment:
My proposed signature for this class is

`Future<TransactionResult> runTransaction(TransactionHandler doTransaction, [ int transactionTimeout = _transactionTimeout])`

Where `TransactionHandler` is defined by `typedef Future<MutableData> TransactionHandler(MutableData data);`

You'll need to define a new class `TransactionResult`:
```
class TransactionResult {
  const TransactionResult._(this.error, this.committed, this.dataSnapshot);
  final DatabaseError error;
  final bool committed;
  final DataSnapshot dataSnapshot;
}
```

Reply:
Done!